rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,253 @@
+
+If any of the following constraints are violated, Vector will error on boot:
+
+- There cannot be several pipelines with the same id (for example `load-balancer.yml` and `load-balancer.json`).
+- The pipeline's configuration files should only contain transforms.
+- A pipeline's transform cannot have the same name as any component from the root configuration.
+- A pipeline cannot use another pipeline's component as input or output.
+
+If the violation occurs during a reload, an error will be triggered and handled in the same fashion as other reload errors.
+
+### Implementation
+
+#### Internal representation
+
+As mentioned in the previous section, a pipeline is _just_ a set of transforms.
+
+To be able to forward the events going through the pipeline to a `sink`, we'll add a new option `outputs` on the pipeline's transforms that will simply specify where the transforms events are redirected to.
+
+`outputs` options are used solely to build the topology and represent an interface between the transform and the external sinks.
+
+A pipeline will have the following internal representation before building the topology.
+
+```rust
+struct PipelineTransform {
+  inner: TransformOuter,
+  outputs: Vec<String>,
+}
+
+struct Pipeline {
+  id: String,
+  transforms: Map<String, PipelineTransform>,
+}
+```
+
+Which corresponds to the following configuration file:
+
+```toml
+# /etc/vector/pipelines/pipeline.toml
+[transforms.foo]
+type = "remap"
+inputs = ["from-root"]
+outputs = ["dc1", "dc2"]
+# ...
+
+[transforms.bar]
+type = "remap"
+inputs = ["foo"]
+outputs = ["dc-us", "dc-eu"]
+# ...
+```
+
+The `outputs` option is made to forward the events from inside the pipeline to an external component.
+
+#### From configuration to topology
+
+If we look deeper at the configuration building process, the configuration compiler will require the pipelines to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+
+To do so, we'll need to implement a `Pipeline` from the previous section. We'll then update [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4) to build a `Config` containing the required pipelines components. The compiler will load the pipelines' transforms and add the `outputs` to the corresponding `sinks`.
+
+The components coming from the pipeline will be cloned inside the final `Config`, in the `IndexMap` containing the `transforms` and the `outputs` from the pipeline components will be added to the referring components input field.
+
+For example, the following configuration and pipeline, and its **equivalent** once built.
+
+```toml
+# /etc/vector/vector.toml
+[sources.in]
+# ...
+
+[sinks.out]
+# ...
+
+# /etc/vector/pipelines/foo.toml
+[transforms.bar]
+inputs = ["in"]
+outputs = ["out"]
```

Comment:
I imagine a components inputs will now be the union of `inputs` and as well as any components where it is listed in `outputs`? Just calling out if we want to still allow sinks to use inputs in addition to being listed as outputs.

Reply:
We don't want to break backward compatibility, so I guess we should keep both.

Reply:
Yeah, I don't think we want to change how things work outside of pipelines. So `outputs` should be a pipelines-exclusive addition and everything else can stay the same.